Please reference README.md for project outline and DATA_DICTIONARY.md for a list of terms. Visuals are available on Tableau public account. All links provided in README.md and on github repo. 

### Part One: Importing and Cleaning Hero Fund Data

In [167]:
import pandas as pd
import requests
import numpy as np

Pay attention to the difference between importing an xlsx file vs a csv file. I had to pip install openpyxl in the terminal. Reading in my first data file.

In [168]:
df = pd.read_excel("C:\\Users\\emorr\\OneDrive\\Desktop\\CHM\\medal.xlsx")
df.dtypes

ACTED        object
LAST         object
FIRST        object
ACTSTATE     object
ACTCODE      object
CITY         object
STRANGER     object
DEATH        object
COUNTY       object
YEAR        float64
dtype: object

I use head() frequently for sanity checks.

In [169]:
df.head()

ACTED      LAST          FIRST ACTSTATE   ACTCODE CITY STRANGER DEATH  \
0   5/6/2021      BAIR        BRANDON       ID        BV  NaN      NaN   NaN   
1  1/13/2022    THOMAS    ADAM LAYMAN       KY  DROWNING  NaN      NaN   NaN   
2  5/17/2021   JOHNSON        ROSS C.       FL  DROWNING  NaN      NaN   NaN   
3  4/11/2021  PETERKIN  JADEN DESHAWN       NC        BB  NaN      NaN   NaN   
4  4/11/2021  PETERKIN        ANTHONY       NC        BB  NaN      NaN   NaN   

  COUNTY    YEAR  
0    NaN  2021.0  
1    NaN  2022.0  
2    NaN  2021.0  
3    NaN  2021.0  
4    NaN  2021.0

Let's drop columns I don't need. The names are not needed and I don't need the full date since I have the year. Use inplace=True to change the actual df

In [170]:
df.drop(columns=['LAST', 'FIRST', 'ACTED'], inplace=True)

In [171]:
df.head()

ACTSTATE   ACTCODE CITY STRANGER DEATH COUNTY    YEAR
0       ID        BV  NaN      NaN   NaN    NaN  2021.0
1       KY  DROWNING  NaN      NaN   NaN    NaN  2022.0
2       FL  DROWNING  NaN      NaN   NaN    NaN  2021.0
3       NC        BB  NaN      NaN   NaN    NaN  2021.0
4       NC        BB  NaN      NaN   NaN    NaN  2021.0

Next I want to change the column names by setting them to an array so I can change them all at once.

In [172]:
df.columns

Index(['ACTSTATE', 'ACTCODE', 'CITY', 'STRANGER', 'DEATH', 'COUNTY', 'YEAR'], dtype='object')

In [173]:
df.columns = ['STATE', 'CODE', 'CITY', 'STRANGER', 'DEATH', 'COUNTY', 'YEAR']

In [174]:
df.head()

STATE      CODE CITY STRANGER DEATH COUNTY    YEAR
0    ID        BV  NaN      NaN   NaN    NaN  2021.0
1    KY  DROWNING  NaN      NaN   NaN    NaN  2022.0
2    FL  DROWNING  NaN      NaN   NaN    NaN  2021.0
3    NC        BB  NaN      NaN   NaN    NaN  2021.0
4    NC        BB  NaN      NaN   NaN    NaN  2021.0

I had to manually input rows for STRANGER, DEATH and COUNTY since this wasn't determined in the data that was provided. I made a judgement call based on each recording of the medal incident if the person was a stranger.  STRANGER with a Y means yes they were strangers. I found the counties based off the cities provided. DEATH with a Y means the medal winner died during the rescue attempt.  I had attempted to write code to scan all of the incident reports, but the language was too broad since they were written between 1904-2022. I hope to revisit this attempt later when I have more experience. I chose KY, CA, and AK for manual entry with the intention of using all three states for this project. However, after multiple attempts I realized I would need to do each state individually. Due to time constraints, this project with focus on California only.

Now I need to filter my data to keep only the rows and columns of CA. I will use the isin function and create a new df to do this.

In [175]:
state = ['CA']
ca_df = df[df['STATE'].isin(state)]
ca_df

STATE         CODE            CITY STRANGER DEATH          COUNTY  \
16       CA     DROWNING       SAN DIEGO        Y     N       SAN DIEGO   
17       CA     DROWNING       SAN DIEGO        Y     N       SAN DIEGO   
27       CA     DROWNING          SANGER        Y     Y          FRESNO   
33       CA       ANIMAL     LOS ANGELES        Y     N              LA   
47       CA     DROWNING       SAN DIEGO        Y     N       SAN DIEGO   
...     ...          ...             ...      ...   ...             ...   
10066    CA  SUFFOCATION         PIONEER        N     N          AMADOR   
10075    CA     DROWNING  SAN BERNARDINO        Y     Y  SAN BERNARDINO   
10076    CA     DROWNING  SAN BERNARDINO        Y     Y  SAN BERNARDINO   
10079    CA     DROWNING         ANTIOCH        Y     N    CONTRA COSTA   
10084    CA     DROWNING       SAN DIEGO        Y     N       SAN DIEGO   

         YEAR  
16     2021.0  
17     2021.0  
27     2020.0  
33     2021.0  
47     2020.0  
...       ...  
10066  1904.0  
10075  1907.0  
10076  1907.0  
10079  1908.0  
10084  1908.0  

[736 rows x 7 columns]

Let's get a feel of what kind of data we have based on type. Let's look at what data is missing and drop what we don't need.

In [176]:
ca_df.dtypes

STATE        object
CODE         object
CITY         object
STRANGER     object
DEATH        object
COUNTY       object
YEAR        float64
dtype: object

In [177]:
ca_df.shape

(736, 7)

In [178]:
ca_df.dropna(inplace=True)

C:\Users\emorr\AppData\Local\Temp\ipykernel_21476\1601242493.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ca_df.dropna(inplace=True)


In [179]:
ca_df.shape

(722, 7)

Do I need to drop anything else so when I get to Part Three and merge my data frames they will be easier to read? I don't think I need City since I will be looking at counties. I might not need years after 2010 since my population data stops at 2010, but I'm going to hang onto them in case I have time to do predictive analysis. 

In [180]:
ca_df.drop(columns=['CITY'], inplace=True)

C:\Users\emorr\AppData\Local\Temp\ipykernel_21476\431470113.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ca_df.drop(columns=['CITY'], inplace=True)


Next I want to get a view of the specifics so when I clean the population data I will have a better idea of what can go and what needs to stay.

In [181]:
ca_df['YEAR'].sort_values(ascending=True).unique()

array([1904., 1907., 1908., 1909., 1910., 1911., 1912., 1913., 1914.,
       1915., 1916., 1917., 1918., 1919., 1920., 1921., 1922., 1923.,
       1924., 1925., 1926., 1927., 1928., 1929., 1930., 1931., 1932.,
       1933., 1934., 1935., 1936., 1937., 1938., 1939., 1940., 1941.,
       1942., 1943., 1944., 1945., 1946., 1947., 1948., 1949., 1950.,
       1951., 1952., 1953., 1954., 1955., 1956., 1957., 1958., 1959.,
       1960., 1961., 1962., 1963., 1964., 1965., 1966., 1967., 1968.,
       1969., 1970., 1971., 1972., 1973., 1974., 1975., 1976., 1977.,
       1978., 1979., 1980., 1981., 1982., 1983., 1984., 1985., 1986.,
       1987., 1988., 1989., 1990., 1991., 1992., 1993., 1994., 1995.,
       1996., 1997., 1998., 1999., 2000., 2001., 2002., 2003., 2004.,
       2005., 2006., 2007., 2008., 2009., 2010., 2011., 2012., 2013.,
       2014., 2015., 2016., 2017., 2018., 2019., 2020., 2021.])

Earliest year 1904, most current year 2021.

In [182]:
ca_df['COUNTY'].unique()

array(['SAN DIEGO', 'FRESNO', 'LA', 'EL DORADO', 'CALAVERAS', 'HUMBOLDT',
       'SAN BERNARDINO', 'RIVERSIDE', 'SACRAMENTO', 'TULARE', 'KERN',
       'SANTA BARBARA', 'ORANGE', 'VENTURA', 'MENDOCINO', 'CONTRA COSTA',
       'STANISLAUS', 'SANTA CRUZ', 'MERCED', 'MARIN', 'SAN JOAQUIN',
       'MONTEREY', 'SANTA CLARA', 'SOLANO', 'ALAMEDA', 'NORTH', 'NEVADA',
       'MARIPOSA', 'INYO', 'PLACER', 'SHASTA', 'TRINITY',
       'SAN LUIS OBISPO', 'SONOMA', 'SAN FRANCISCO', 'SISKIYOU', 'MONO',
       'BUTTE', 'COLUSA', 'SAN MATEO', 'TEHAMA', 'YOLO', 'YUBA', 'LAKE',
       'IMERIAL', 'IMPERIAL', 'KINGS', 'DEL NORTE', 'MADERA', 'TUOLUMNE',
       'SUTTER', 'NAPA', 'HUMBOLDT ', 'LAKE ', 'GLENN ', 'LASSEN',
       'SAN MATEO ', 'AMADOR'], dtype=object)

Awarded in 58 counties

In [183]:
No = ca_df['STRANGER'] == 'N'
ca_df.loc[No, ['COUNTY','YEAR', 'STRANGER']].value_counts()

COUNTY         YEAR    STRANGER
LA             2011.0  N           4
SAN FRANCISCO  1913.0  N           3
LA             1959.0  N           3
SAN MATEO      2000.0  N           3
SAN DIEGO      1965.0  N           3
                                  ..
MARIN          1914.0  N           1
               1923.0  N           1
               1963.0  N           1
               1966.0  N           1
YUBA           1989.0  N           1
Length: 209, dtype: int64

209 of the medal recipients were not strangers.

In [184]:
Yes = ca_df['STRANGER'] == 'Y'
ca_df.loc[Yes, ['COUNTY','YEAR', 'STRANGER']].value_counts()

COUNTY   YEAR    STRANGER
PLACER   1970.0  Y           5
LA       1980.0  Y           5
VENTURA  1987.0  Y           5
LA       1964.0  Y           4
         1967.0  Y           3
                            ..
ORANGE   1963.0  Y           1
         1966.0  Y           1
         1967.0  Y           1
         1970.0  Y           1
YUBA     1997.0  Y           1
Length: 347, dtype: int64

347 of medal recipients were strangers. 

In [185]:
ca_df['CODE'].value_counts() 

DROWNING         234
BV               161
BB                80
ASSAULT           70
SUFFOCATION       37
MVP               35
ELEVATION         22
ELECTROCUTION     14
GENERAL           13
ANIMAL            13
SV                12
BOAT              12
ICE                5
CAVE-IN            5
IMPENDING EX.      5
MVI                3
MVV                1
Name: CODE, dtype: int64

I think I have the medal data where I want it. Let's move on to reading in the California county population data from the census. 

### Part Two: Importing and Cleaning California Population Data

Reading in my second data file.

In [186]:
pop = pd.read_excel("C:\\Users\\emorr\\OneDrive\\Desktop\\CHM\\CAPOP.xlsx")


In [187]:
pop

COUNTY  Unnamed: 1  INCORP  Unnamed: 3  YEAR 1850  YEAR 1860  \
0           Alameda         NaN    1853         NaN        NaN     8927.0   
1            Alpine         NaN    1864         NaN        NaN      685.0   
2            Amador         NaN    1854         NaN    10930.0     9582.0   
3             Butte         NaN    1850         NaN    12106.0    11403.0   
4         Calaveras         NaN    1850         NaN    16299.0     8895.0   
5            Colusa         NaN    1850         NaN     2274.0     6165.0   
6      Contra Costa         NaN    1850         NaN     5328.0     8461.0   
7         Del Norte         NaN    1857         NaN     1993.0     2022.0   
8         El Dorado         NaN    1850         NaN    20562.0    10309.0   
9            Fresno         NaN    1856         NaN     4605.0     6336.0   
10            Glenn         NaN    1891         NaN        NaN        NaN   
11         Humboldt         NaN    1853         NaN     2694.0     6140.0   
12         Imperial         NaN    1907         NaN        NaN        NaN   
13             Inyo         NaN    1866         NaN        NaN     1956.0   
14             Kern         NaN    1866         NaN        NaN     2925.0   
15            Kings         NaN    1893         NaN        NaN        NaN   
16             Lake         NaN    1861         NaN        NaN     2969.0   
17           Lassen         NaN    1864         NaN        NaN     1327.0   
18      Los Angeles         NaN    1850         NaN    11333.0    15309.0   
19           Madera         NaN    1893         NaN        NaN        NaN   
20            Marin         NaN    1850         NaN     3334.0     6903.0   
21         Mariposa         NaN    1850         NaN     6243.0     4572.0   
22        Mendocino         NaN    1850         NaN     3967.0     7545.0   
23           Merced         NaN    1855         NaN     1141.0     2807.0   
24            Modoc         NaN    1874         NaN        NaN        NaN   
25             Mono         NaN    1861         NaN        NaN      430.0   
26         Monterey         NaN    1850         NaN     4739.0     9876.0   
27             Napa         NaN    1850         NaN     5521.0     7163.0   
28           Nevada         NaN    1851         NaN    16446.0    19134.0   
29           Orange         NaN    1889         NaN        NaN        NaN   
30           Placer         NaN    1851         NaN    13270.0    11357.0   
31           Plumas         NaN    1854         NaN     4363.0     4489.0   
32        Riverside         NaN    1893         NaN        NaN        NaN   
33       Sacramento         NaN    1850         NaN    24142.0    26830.0   
34       San Benito         NaN    1874         NaN        NaN        NaN   
35   San Bernardino         NaN    1853         NaN     5551.0     3988.0   
36        San Diego         NaN    1850         NaN     4324.0     4951.0   
37    San Francisco         NaN    1850         NaN    56802.0   149473.0   
38      San Joaquin         NaN    1850         NaN     9435.0    21050.0   
39  San Luis Obispo         NaN    1850         NaN     1782.0     4772.0   
40        San Mateo         NaN    1856         NaN     3214.0     6635.0   
41    Santa Barbara         NaN    1850         NaN     3543.0     7784.0   
42      Santa Clara         NaN    1850         NaN    11912.0    26246.0   
43       Santa Cruz         NaN    1850         NaN     4944.0     8743.0   
44           Shasta         NaN    1850         NaN     4360.0     4173.0   
45           Sierra         NaN    1852         NaN    11387.0     5619.0   
46         Siskiyou         NaN    1852         NaN     7629.0     6848.0   
47           Solano         NaN    1850         NaN     7169.0    16871.0   
48           Sonoma         NaN    1850         NaN    11867.0    19819.0   
49       Stanislaus         NaN    1854         NaN     2245.0     6499.0   
50           Sutter         NaN    1850         NaN     3390.0     5030.0   
51       

Right away I see that 2 columns are totally blank. Let's get rid of those. (keep in mind spacing when using the column names)

In [188]:
pop.drop(columns=['Unnamed: 1', 'Unnamed: 3'], inplace=True)

In [189]:
pop.head()

COUNTY  INCORP  YEAR 1850  YEAR 1860  YEAR 1870  YEAR 1880  YEAR 1890  \
0    Alameda    1853        NaN     8927.0    24237.0    62976.0    93864.0   
1     Alpine    1864        NaN      685.0      539.0      667.0      509.0   
2     Amador    1854    10930.0     9582.0    11384.0    10320.0    11116.0   
3      Butte    1850    12106.0    11403.0    18721.0    17939.0    17117.0   
4  Calaveras    1850    16299.0     8895.0     9094.0     8882.0    11200.0   

   YEAR 1900  YEAR 1910  YEAR 1920  YEAR 1930  YEAR 1940  YEAR 1950  \
0     130197     246131     344177     474883     513011     740315   
1        309        243        241        323        241        397   
2       9086       7793       8494       8973       9151       9990   
3      27301      30030      34093      42840      64930      82030   
4       9171       6183       6008       8221       9902      10289   

   YEAR 1960  YEAR 1970  YEAR 1980  YEAR 1990  YEAR 2000  YEAR 2010  
0     908209    1071446    1105379    1279182    1443741    1510271  
1        484       1097       1113       1208       1175       1204  
2      11821      19314      30039      35100      38091      40474  
3     101969     143851     182120     203171     220000     211632  
4      13585      20710      31998      40554      45578      45292

Let's see if the counties in both dataframes match. I think I may use County to merge the  dataframes in Part Three.

In [190]:
#be aware this is medal data
ca_df['COUNTY'].unique()

array(['SAN DIEGO', 'FRESNO', 'LA', 'EL DORADO', 'CALAVERAS', 'HUMBOLDT',
       'SAN BERNARDINO', 'RIVERSIDE', 'SACRAMENTO', 'TULARE', 'KERN',
       'SANTA BARBARA', 'ORANGE', 'VENTURA', 'MENDOCINO', 'CONTRA COSTA',
       'STANISLAUS', 'SANTA CRUZ', 'MERCED', 'MARIN', 'SAN JOAQUIN',
       'MONTEREY', 'SANTA CLARA', 'SOLANO', 'ALAMEDA', 'NORTH', 'NEVADA',
       'MARIPOSA', 'INYO', 'PLACER', 'SHASTA', 'TRINITY',
       'SAN LUIS OBISPO', 'SONOMA', 'SAN FRANCISCO', 'SISKIYOU', 'MONO',
       'BUTTE', 'COLUSA', 'SAN MATEO', 'TEHAMA', 'YOLO', 'YUBA', 'LAKE',
       'IMERIAL', 'IMPERIAL', 'KINGS', 'DEL NORTE', 'MADERA', 'TUOLUMNE',
       'SUTTER', 'NAPA', 'HUMBOLDT ', 'LAKE ', 'GLENN ', 'LASSEN',
       'SAN MATEO ', 'AMADOR'], dtype=object)

In [191]:
pop['COUNTY'].unique()

array(['Alameda', 'Alpine', 'Amador', 'Butte', 'Calaveras', 'Colusa',
       'Contra Costa', 'Del Norte', 'El Dorado', 'Fresno', 'Glenn',
       'Humboldt', 'Imperial', 'Inyo', 'Kern', 'Kings', 'Lake', 'Lassen',
       'Los Angeles', 'Madera', 'Marin', 'Mariposa', 'Mendocino',
       'Merced', 'Modoc', 'Mono', 'Monterey', 'Napa', 'Nevada', 'Orange',
       'Placer', 'Plumas', 'Riverside', 'Sacramento', 'San Benito',
       'San Bernardino', 'San Diego', 'San Francisco', 'San Joaquin',
       'San Luis Obispo', 'San Mateo', 'Santa Barbara', 'Santa Clara',
       'Santa Cruz', 'Shasta', 'Sierra', 'Siskiyou', 'Solano', 'Sonoma',
       'Stanislaus', 'Sutter', 'Tehama', 'Trinity', 'Tulare', 'Tuolumne',
       'Ventura', 'Yolo', 'Yuba'], dtype=object)

Make all the counties upper case.

In [192]:
pop['COUNTY'] = pop['COUNTY'].str.upper()

In [193]:
pop.COUNTY.unique()

array(['ALAMEDA', 'ALPINE', 'AMADOR', 'BUTTE', 'CALAVERAS', 'COLUSA',
       'CONTRA COSTA', 'DEL NORTE', 'EL DORADO', 'FRESNO', 'GLENN',
       'HUMBOLDT', 'IMPERIAL', 'INYO', 'KERN', 'KINGS', 'LAKE', 'LASSEN',
       'LOS ANGELES', 'MADERA', 'MARIN', 'MARIPOSA', 'MENDOCINO',
       'MERCED', 'MODOC', 'MONO', 'MONTEREY', 'NAPA', 'NEVADA', 'ORANGE',
       'PLACER', 'PLUMAS', 'RIVERSIDE', 'SACRAMENTO', 'SAN BENITO',
       'SAN BERNARDINO', 'SAN DIEGO', 'SAN FRANCISCO', 'SAN JOAQUIN',
       'SAN LUIS OBISPO', 'SAN MATEO', 'SANTA BARBARA', 'SANTA CLARA',
       'SANTA CRUZ', 'SHASTA', 'SIERRA', 'SISKIYOU', 'SOLANO', 'SONOMA',
       'STANISLAUS', 'SUTTER', 'TEHAMA', 'TRINITY', 'TULARE', 'TUOLUMNE',
       'VENTURA', 'YOLO', 'YUBA'], dtype=object)

That worked. It looks like the only spelling difference between the data frames is Los Angeles so let's change that with the at() function.

In [194]:
pop.at[18, 'COUNTY'] = 'LA'

In [195]:
pop.COUNTY.unique()

array(['ALAMEDA', 'ALPINE', 'AMADOR', 'BUTTE', 'CALAVERAS', 'COLUSA',
       'CONTRA COSTA', 'DEL NORTE', 'EL DORADO', 'FRESNO', 'GLENN',
       'HUMBOLDT', 'IMPERIAL', 'INYO', 'KERN', 'KINGS', 'LAKE', 'LASSEN',
       'LA', 'MADERA', 'MARIN', 'MARIPOSA', 'MENDOCINO', 'MERCED',
       'MODOC', 'MONO', 'MONTEREY', 'NAPA', 'NEVADA', 'ORANGE', 'PLACER',
       'PLUMAS', 'RIVERSIDE', 'SACRAMENTO', 'SAN BENITO',
       'SAN BERNARDINO', 'SAN DIEGO', 'SAN FRANCISCO', 'SAN JOAQUIN',
       'SAN LUIS OBISPO', 'SAN MATEO', 'SANTA BARBARA', 'SANTA CLARA',
       'SANTA CRUZ', 'SHASTA', 'SIERRA', 'SISKIYOU', 'SOLANO', 'SONOMA',
       'STANISLAUS', 'SUTTER', 'TEHAMA', 'TRINITY', 'TULARE', 'TUOLUMNE',
       'VENTURA', 'YOLO', 'YUBA'], dtype=object)

Since our medal data doesn't start until 1904, let's get rid of the columns with years prior to 1904. I also don't need the year incorporated.

In [196]:
pop.drop(columns=['INCORP', 'YEAR 1850', 'YEAR 1860', 'YEAR 1870', 'YEAR 1880', 'YEAR 1890'], inplace=True)

In [197]:
pop.shape

(58, 13)

In [198]:
pop.dtypes

COUNTY       object
YEAR 1900     int64
YEAR 1910     int64
YEAR 1920     int64
YEAR 1930     int64
YEAR 1940     int64
YEAR 1950     int64
YEAR 1960     int64
YEAR 1970     int64
YEAR 1980     int64
YEAR 1990     int64
YEAR 2000     int64
YEAR 2010     int64
dtype: object

The population data seems in good shape. 

### Part Three: Merging the Data

First let's check our columns for each.

In [199]:
ca_df.columns

Index(['STATE', 'CODE', 'STRANGER', 'DEATH', 'COUNTY', 'YEAR'], dtype='object')

In [200]:
pop.columns

Index(['COUNTY', 'YEAR 1900', 'YEAR 1910', 'YEAR 1920', 'YEAR 1930',
       'YEAR 1940', 'YEAR 1950', 'YEAR 1960', 'YEAR 1970', 'YEAR 1980',
       'YEAR 1990', 'YEAR 2000', 'YEAR 2010'],
      dtype='object')

Now let's merge. By merging on the County I will be able to create a new value where the year of the incident will be right there with the population of the corresponding decade in that county. I will use this to make a new column shown in the next few steps.

In [201]:
medal_pop = pd.merge(ca_df, pop)
medal_pop.columns

Index(['STATE', 'CODE', 'STRANGER', 'DEATH', 'COUNTY', 'YEAR', 'YEAR 1900',
       'YEAR 1910', 'YEAR 1920', 'YEAR 1930', 'YEAR 1940', 'YEAR 1950',
       'YEAR 1960', 'YEAR 1970', 'YEAR 1980', 'YEAR 1990', 'YEAR 2000',
       'YEAR 2010'],
      dtype='object')

Sanity check.

In [202]:
medal_pop.head()

STATE      CODE STRANGER DEATH     COUNTY    YEAR  YEAR 1900  YEAR 1910  \
0    CA  DROWNING        Y     N  SAN DIEGO  2021.0      61665     112248   
1    CA  DROWNING        Y     N  SAN DIEGO  2021.0      61665     112248   
2    CA  DROWNING        Y     N  SAN DIEGO  2020.0      61665     112248   
3    CA  DROWNING        Y     N  SAN DIEGO  2020.0      61665     112248   
4    CA   ASSAULT        Y     N  SAN DIEGO  2019.0      61665     112248   

   YEAR 1920  YEAR 1930  YEAR 1940  YEAR 1950  YEAR 1960  YEAR 1970  \
0     209659     289348     556808    1033011    1357854    1861846   
1     209659     289348     556808    1033011    1357854    1861846   
2     209659     289348     556808    1033011    1357854    1861846   
3     209659     289348     556808    1033011    1357854    1861846   
4     209659     289348     556808    1033011    1357854    1861846   

   YEAR 1980  YEAR 1990  YEAR 2000  YEAR 2010  
0    2498016    2813833    3095313    3298634  
1    2498016    2813833    3095313    3298634  
2    2498016    2813833    3095313    3298634  
3    2498016    2813833    3095313    3298634  
4    2498016    2813833    3095313    3298634

This new column will help my visuals tell a more concise story. It will be easier to see the population of the county for that incident year and the decade group it belongs to.

In [203]:
conditions = [(medal_pop['YEAR'] <= 1910.0),
    (medal_pop['YEAR'] >= 1911.0) & (medal_pop['YEAR']<=1920.0),
    (medal_pop['YEAR'] >= 1921.0) & (medal_pop['YEAR']<=1930.0),
    (medal_pop['YEAR'] >= 1931.0) & (medal_pop['YEAR']<=1940.0),
    (medal_pop['YEAR'] >= 1941.0) & (medal_pop['YEAR']<=1950.0),
    (medal_pop['YEAR'] >= 1951.0) & (medal_pop['YEAR']<=1960.0),
    (medal_pop['YEAR'] >= 1961.0) & (medal_pop['YEAR']<=1970.0),
    (medal_pop['YEAR'] >= 1971.0) & (medal_pop['YEAR']<=1980.0),
    (medal_pop['YEAR'] >= 1981.0) & (medal_pop['YEAR']<=1990.0),
    (medal_pop['YEAR'] >= 1991.0) & (medal_pop['YEAR']<=2000.0),
    (medal_pop['YEAR'] >= 2001.0) & (medal_pop['YEAR']<=2010.0),
    (medal_pop['YEAR'] >= 2011.0) & (medal_pop['YEAR']<=2020.0),
    ]


values = ['Year 1900', 'Year 1910','Year 1920','Year 1930','Year 1940', 'Year 1950', 'Year 1960','Year 1970','Year 1980', 'Year 1990', 'Year 2000', 'Year 2010']

medal_pop['DECADE'] = np.select(conditions, values)

medal_pop

STATE         CODE STRANGER DEATH     COUNTY    YEAR  YEAR 1900  \
0      CA     DROWNING        Y     N  SAN DIEGO  2021.0      61665   
1      CA     DROWNING        Y     N  SAN DIEGO  2021.0      61665   
2      CA     DROWNING        Y     N  SAN DIEGO  2020.0      61665   
3      CA     DROWNING        Y     N  SAN DIEGO  2020.0      61665   
4      CA      ASSAULT        Y     N  SAN DIEGO  2019.0      61665   
..    ...          ...      ...   ...        ...     ...        ...   
710    CA     DROWNING        Y     N     SUTTER  1965.0       6328   
711    CA     DROWNING        N     N       NAPA  1962.0      19800   
712    CA     DROWNING        N     N       NAPA  1962.0      19800   
713    CA           BB        N     N     LASSEN  1913.0       4802   
714    CA  SUFFOCATION        N     N     AMADOR  1904.0       9086   

     YEAR 1910  YEAR 1920  YEAR 1930  YEAR 1940  YEAR 1950  YEAR 1960  \
0       112248     209659     289348     556808    1033011    1357854   
1       112248     209659     289348     556808    1033011    1357854   
2       112248     209659     289348     556808    1033011    1357854   
3       112248     209659     289348     556808    1033011    1357854   
4       112248     209659     289348     556808    1033011    1357854   
..         ...        ...        ...        ...        ...        ...   
710      10115      14618      18680      26239      33380      41935   
711      20678      22897      28503      46603      65890      79140   
712      20678      22897      28503      46603      65890      79140   
713       8507      12589      14479      18474      13597      16796   
714       7793       8494       8973       9151       9990      11821   

     YEAR 1970  YEAR 1980  YEAR 1990  YEAR 2000  YEAR 2010     DECADE  
0      1861846    2498016    2813833    3095313    3298634          0  
1      1861846    2498016    2813833    3095313    3298634          0  
2      1861846    2498016    2813833    3095313    3298634  Year 2010  
3      1861846    2498016    2813833    3095313    3298634  Year 2010  
4      1861846    2498016    2813833    3095313    3298634  Year 2010  
..         ...        ...        ...        ...        ...        ...  
710      52246      64415      78930      94737      99633  Year 1960  
711      99199     110765     124279     136484     138019  Year 1960  
712      99199     110765     124279     136484     138019  Year 1960  
713      21661      27598      33828      34895      32730  Year 1910  
714      19314      30039      35100      38091      40474  Year 1900  

[715 rows x 19 columns]

Now I can go through and see the population of the county for the year of that act. For example, the next code snippet shows that all acts from 1900-1910 in the highest population county, LA (504,131) there were 5 awards with 2 being strangers. Whereas the smallest population county Lassen (4,802) there was 1 award of a nonstranger. It's worth noting that the counties with the awards for strangers are higher in counties that aren't low population. 

In [204]:
nineteen = medal_pop['DECADE'] == 'Year 1910' 
medal_pop.loc[nineteen, ['COUNTY','YEAR 1900', 'STRANGER']].value_counts()

COUNTY          YEAR 1900  STRANGER
LA              504131     N           3
HUMBOLDT        33857      N           3
SAN FRANCISCO   416912     N           3
RIVERSIDE       34696      N           2
SAN DIEGO       61665      Y           2
LA              504131     Y           2
SANTA CRUZ      26140      Y           2
VENTURA         18347      N           1
SANTA CRUZ      26140      N           1
SANTA BARBARA   27738      N           1
ALAMEDA         130197     Y           1
SAN BERNARDINO  56706      N           1
ORANGE          34436      Y           1
                           N           1
MARIN           25114      N           1
LASSEN          4802       N           1
YOLO            13926      N           1
dtype: int64

Let's look at the awards given between 1981-1990 to see if much has changed in 70ish years.

In [205]:
nineteeneighty = medal_pop['DECADE'] == 'Year 1980' 
medal_pop.loc[nineteeneighty, ['COUNTY','YEAR 1980', 'STRANGER']].value_counts()

COUNTY          YEAR 1980  STRANGER
LA              8863164    Y           15
SAN DIEGO       2498016    Y            8
VENTURA         669016     Y            8
SAN BERNARDINO  1418380    Y            7
ORANGE          2410556    Y            7
SANTA CLARA     1497577    Y            4
ALAMEDA         1105379    Y            4
SONOMA          388222     Y            4
SAN FRANCISCO   723959     Y            3
SAN BERNARDINO  1418380    N            3
SACRAMENTO      1041219    Y            3
RIVERSIDE       1170413    Y            3
SANTA CRUZ      229734     Y            3
LA              8863164    N            3
ORANGE          2410556    N            2
KERN            543477     N            2
YUBA            58228      N            2
SAN DIEGO       2498016    N            2
CONTRA COSTA    803732     Y            2
SONOMA          388222     N            2
CONTRA COSTA    803732     N            2
SISKIYOU        43531      Y            2
MONTEREY        355660     N            

Let's look at only the number of medal recipients by county.

In [206]:
medal_pop.COUNTY.value_counts()

LA                 128
SAN DIEGO           53
ORANGE              52
SAN FRANCISCO       37
RIVERSIDE           27
SAN BERNARDINO      24
ALAMEDA             22
CONTRA COSTA        22
MONTEREY            21
SAN MATEO           21
VENTURA             21
SACRAMENTO          21
SONOMA              20
SANTA CRUZ          18
HUMBOLDT            18
SANTA CLARA         15
STANISLAUS          14
SANTA BARBARA       13
KERN                13
TULARE              12
PLACER              11
DEL NORTE            9
MENDOCINO            9
FRESNO               9
MARIPOSA             8
MERCED               8
MARIN                8
SAN JOAQUIN          7
EL DORADO            6
YUBA                 6
SOLANO               6
SISKIYOU             6
NEVADA               5
TUOLUMNE             5
SAN LUIS OBISPO      5
SHASTA               5
MONO                 4
BUTTE                3
SUTTER               2
COLUSA               2
TRINITY              2
TEHAMA               2
YOLO                 2
IMPERIAL   

Before the final part of this project check out the Tableau dashboard and story.

Now I want to save my medal_pop data frame as a new xlsx file. Uing the ExcelWriter() method. I'm doing this so I can upload the data in Tableau.

In [207]:
datatoexcel = pd.ExcelWriter('ProjectHero.xlsx')

medal_pop.to_excel(datatoexcel)

datatoexcel.save()


### Part Four: Findings

Is a person more likey to risk her or his life to save the life of a stranger, if the act is taking place in a smaller community?    This question stemmed from something I read; the author theorized this to be true based on an unconscious assumption of small town folks that the 'stranger' is a genetic relation unknown. 
<br>
Based on the data I used, I found that population size does not dictate the likelihood of a stranger attempting to save your life.  There were 138 more strangers awarded than nonstrangers across all population sizes.    Of course larger population areas offer more opportunities for someone being in need and being a stranger. 
<br>
<br>
*Interesting thoughts and findings to take with you:*
